src/cell_group.hpp

```diff
@@ -143,51 +143,61 @@ class cell_group {
     }
 
     void advance(time_type tfinal, time_type dt) {
-        while (lowered_.time()<tfinal) {
-            // take any pending samples
-            time_type cell_time = lowered_.time();
+        EXPECTS(lowered_.state_synchronized());
```

Comment:
Ignoring the extra comments, the function is pretty much the same length as before.

That said, it will be much simpler when sampling is also pushed to the back-end, so I would defer a refactoring of this method until then.